File: src/main/java/io/springacademy/shortener/LinkService.java
package io.springacademy.shortener;

import org.apache.commons.lang3.RandomStringUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import reactor.core.publisher.Mono;

@Service
public class LinkService {

    private final String domain;
    private final LinkRepository linkRepository;

    public LinkService(@Value("${domain}") String domain, LinkRepository linkRepository) {
        this.domain = domain;
        this.linkRepository = linkRepository;
    }

    Mono<String> shortenLink(String link) {
        String randomKey = RandomStringUtils.randomAlphabetic(8);
        return linkRepository.save(new Link(randomKey, link))
                             .map(savedLink -> domain + "/" + randomKey);
    }

    Mono<String> findOriginalLink(String key) {
        return linkRepository.findById(key)
                             .map(Link::getOriginalLink);
    }
}

File: src/main/java/io/springacademy/shortener/LinkService.kt
package io.springacademy.shortener

import org.apache.commons.lang3.RandomStringUtils
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

import reactor.core.publisher.Mono

@Service
class LinkService(
    @Value("\${domain}") private val domain: String,
    private val linkRepository: LinkRepository
) {

    fun shortenLink(link: String): Mono<String> {
        val randomKey = RandomStringUtils.randomAlphabetic(8)
        return linkRepository.save(Link(randomKey, link))
            .map { "$domain/$randomKey" }
    }

    fun findOriginalLink(key: String): Mono<String> {
        return linkRepository.findById(key)
            .map { it.originalLink }
    }
}
